### ISSUE TITLE
Font Editor for Ink/Stitch

### ISSUE THREAD

**Issue by Augusa** [inkstitch/inkstitch#586]
I created a font from ["Déjà vu"](https://www.fontsquirrel.com/fonts/dejavu-serif) 
I try to use the font editor of inkscape. But I am not able to do anything with my letters. Is there a secret? I read that Inkstitch fonts are made with. So I see it is possible but how?

**Comment by danielkschneider** [comment]
Hi guys
Let me rephrase this. She created a font that would be quite interesting to have. You can see it in action here: http://lyogau.over-blog.com/2019/12/sacs-a-chaussure-en-chanson.html

Basically, she would like to know how to contribute it to InkStitch and that raises a few questions:
- how to use to InkScape font editor. If my understanding is right, it does the right baseline positioning and also provides a way to name and organize a layer for each glyph. The glyphs (that sit all in the same position) then do not need to be exported to a "real" font.
- how to make Ink/Stitch compatible glyphs. IMHO there shouldn't be anything special, except making the parts satin-compatible (either equal points or rungs) and put them in the right order.
- how to package these for Ink/Stitch. If I understand right, there are four versions (one for each writing direction) plus a JSON file that is fairly self-explaining. But it would be nice if could shortly say what each JSON property does.

To summarize: If someone had time, it would be nice to have some documentation that explains how to contribute a font, or at least complete/correct misunderstandings I made above. As for her precise question above, I think I found an answer, i.e. a good french documentation on how to create fonts: https://fr.flossmanuals.net/fontes-libres/svg-fonts-dans-inkscape/

Cheers and happy Xmas and hopefully vacation to all ! Daniel

**Comment by lexelby** [comment]
Wow, that is a beautiful font!  Documenting how to create Ink/Stitch fonts is definitely on my list.  The current state of things is pretty rudimentary to say the least.

I used the Inkscape glyph editing extensions, but you definitely don't have to.  For now, probably the easiest way to proceed would be to copy an existing font like inkstitch-small and replace all the glyphs with your own.  Do the left-to-right one first, then copy that file to the other three and adjust the order of the satins as necessary.

Inkstitch-small shows an example of how to adjust kerning, but it only has a single glyph pair (w followed by o).  Add more entries in that JSON object to adjust the kerning for other glyph pairs.

I dream of making a full UI for font editing someday.

**Comment by lexelby** [comment]
One more thing: just make sure that we comply with the licensing requirements for Deja Vu.  Since the Deja Vu modifications to Bitstream Vera are public domain, it looks like all we need to do is include the Bitstream Vera license itself.

**Comment by Augusa** [comment]
Thanks for the answers. I try to use it soon.

**Comment by Augusa** [comment]
I am trying to place my glyphs in the left-to-right svg. My letters need a bigger canvas than small font. I create them with 100x100px and my guidelines are differents from yours. Is it possible to change them for exemple in tt-directors. Your glyphs seem to be placed in the middle of the Square on the baseline. Is it important? It's difficult for me to fix the kerning without seeing the result. Is it possible to test the font? I create letters with accent. I need other glyphlayers. How can I do? All this questions are very technical for me. 
I am also curious to understand how you create your svg file with glyphlayers with Inkscape. I never succeed to enter my glyphs for embroidery in Inkscape font editor. You say somewhere you do it with, I should be happy to discover how it is possible.

**Comment by lexelby** [comment]
I'm sorry this is such a mess!  This is definitely a work in progress.

Don't use the Inkscape builtin font editor.  There's an extension in the extensions menu under Typography that can create a glyph layer for you.   That's how I do it.

Ink/Stitch only pays attention to the baseline guide.  It ensures that all letters line up on the same baseline.  You can use the other guides to help you ensure that your letters match up with each other.

Kerning is very much an afterthought currently.  I would recommend you open the Lettering extension and type in something like this: aaabacadaeafafahaiajak (etc) to test out all the kerning pairs.  I hope to make this easier to do down the road!

**Comment by Augusa** [comment]
@lexelby, Thank you very much for your help. I have discovered the extension Typography. So, I am ready to start a correct file. Of course I have another question. The deja Vu font seams to have a different spacing for capital letters and lower case. Is it possible?
I am working on the kerning. I use the font to write in Inkscape and I try to do the same spacing for the embroidered letters. I start to see some groups of letters which are running in the same way. But I think I need 100 years to finish!!
Is there deadline to finish for exemple for a new version? 
My next work will be probably the bold font. I found it's not so terrible when the regular is created. But first of all to finish the kerning.  It exists surely somewhere in a file but I don't find it.

**Comment by Augusa** [comment]
@lexelby, My files with glyphs are ready. Does exist a way to test them?
I prepare also the kerning. I have a small question. What is the sign between the elements for each pair of letters. As you give only 1 exemple in your json file I am not sure.

**Comment by lexelby** [comment]
> The deja Vu font seams to have a different spacing for capital letters and lower case. Is it possible?

Not yet -- I didn't even realize that was a possibility for fonts.  How does this work?  Is the spacing wider if you have two capital letters in a row?  What about a capital letter followed by a lower-case letter?  What about lower-case followed by capital?  I wonder if they just did this using kerning...

> But I think I need 100 years to finish!!

I've always wondered how font designers avoid going insane...

> Is there deadline to finish for exemple for a new version?

Definitely not.  As you can tell, development has been pretty slow lately.  In any case, I would be happy to create a release just for your new font.

> Does exist a way to test them?

The best way to do it would be to push a new branch with your font in it.  You should already have write access to this repository if I added you correctly.  Once you push a branch, a build will automatically start and create a development release that we can all download and test.

Now that I think about it windows building on Travis is broken.  You'll need to base your branch off of my `lneva/stdout-encoding` branch so that you get the new GitHub Actions build system.  So something like this:

```
$ git fetch origin
$ git checkout -b augusa/deja-vu
```

Then commit your new font directory to that branch and push.

> I prepare also the kerning. I have a small question. What is the sign between the elements for each pair of letters. As you give only 1 exemple in your json file I am not sure.

Sorry about that!  I'd always intended to go back and document this but I haven't had a chance yet.  You're blazing a new trail, and I really appreciate your bravery. :)

The kerning number is the number of millimeters of spacing to _add_ between the letters.  For example, if your font.json looks like this:

```
  "letter_spacing": 1.5,
  "kerning_pairs": {
    "ab": 0.2
  }
```

That tells Ink/Stitch that an "a" followed by a "b" should have 1.7mm of spacing between them.  Negative numbers move the letters closer together.

**Comment by Augusa** [comment]
@lexelby I didn't see your answer. Thank you.
I am completly unable to create a branch on github. Is it possible to send you my file of glyph somewhere.
For the kerning my question is only how to separate the pairs of letters in the json file. Your exemple has only one pair and I don't know if I have to separate the pairs with "," or with ";" or something else. 
You are right, the work on kerning is something mad! I prepared a lot of pairs with particular distance but now I am affraid I probably made errors in such an amont of things. What is the best? Use the letters without the kerning and everybody fixes the lettering or using the file with kerning  with probable errors?
I will be very happy if I can help a little Inkstitch. I enjoy really to work with even if I miss a lot of possibilities I can find in Stitch Era. It's a challenge for me to find solutions with inkscape to get interessant results. Thanks for your Inkstitch

**Comment by kaalleen** [comment]
@Augusa 
The separation of the kerning pairs need an comma (,).
Example:

```
  "letter_spacing": 1.5,
  "kerning_pairs": {
    "ab": 0.2,
    "ac": 0.1,
    "ad": 0.3
  }
```

Hope this helps.

**Comment by lexelby** [comment]
@Augusa If you just post your files here as a .ZIP, we can take care of creating the branch for you, no problem.  Thank you for your work!

 I think trying to set the kerning is going to be very difficult for you right now because you don't have a development environment for Ink/Stitch set up.  We need to make a font development mode that notices changes in the JSON file and automatically re-reads it so that you can test kerning on the fly.  Without that, you'll have a _very_ slow process of trial and error, I think...

So I guess what I'm saying is, let's leave it as it is now and people can manually fix the kerning.  Once we create a font development mode, you can come back and do the kerning.

Sound good?

**Comment by Augusa** [comment]
@lexelby Now I add my two files ( glyphfile and jsonfile) I am not able to do anything else with them. The question of licence is not solved. I put my name in the properties of glyphfile but I am not able to decide what must be written or not. My jsonfile is just a try. I hope you can do something with my work. Thanks for your help
[dejavufont.zip](https://github.com/inkstitch/inkstitch/files/4318927/dejavufont.zip)

**Comment by LochanLoch** [comment]
happy to find some help in closed issue :-)
